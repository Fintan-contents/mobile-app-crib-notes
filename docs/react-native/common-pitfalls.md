---
title: ハマりがちな落とし穴
---

React Nativeで開発する際に起きた困り事や不具合についてまとめています。
