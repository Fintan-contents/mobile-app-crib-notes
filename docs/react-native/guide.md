---
title: 何らかのガイドを書きたい気持ちはある
---
