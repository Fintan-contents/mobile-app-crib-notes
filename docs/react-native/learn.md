---
title: About
hide_table_of_contents: true
---

このコンテンツはReact Nativeでのモバイルアプリケーション開発を学習するためのコンテンツです。

ReactでWebアプリケーションを実装できるがReact Nativeでモバイルアプリケーションを実装したことがないアプリケーション開発者を対象として、次のことについて学ぶことを目的としています。

- ReactとReact Nativeの違い
- React Nativeで作成したモバイルアプリケーションの動作確認方法

また、作成したモバイルアプリケーションの配布方法やテスト手法についてはこのコンテンツでは触れません。

<!-- textlint-disable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->

import {PageList} from '../../src/components';
const overviews = [
  {
    title: 'Getting Started',
    to: '/react-native/learn/getting-started',
    summary: <>React Nativeでのアプリ開発に必要な環境を用意します。</>
  },
  {
    title: 'アプリの実装前に',
    to: '/react-native/learn/basic-concepts',
    summary: <>React Nativeでアプリを実装し始める前に最低限必要となる前提知識や、React NativeとReact Navigationの基本的な知識について学びます。</>
  },
  {
    title: 'ToDoアプリの実装',
    to: '/react-native/learn/todo-app',
    summary: <>React Nativeで簡単なToDoアプリを実装します。</>
  },
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->
