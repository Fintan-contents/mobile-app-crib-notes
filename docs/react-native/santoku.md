---
title: Example Application
sidebar_label: About
hide_table_of_contents: true
---

React Nativeサンプルアプリケーションの設計ドキュメントです。

<!-- textlint-disable ja-technical-writing/sentence-length -->

import {PageList} from '../../src/components';
const overviews = [
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length -->
