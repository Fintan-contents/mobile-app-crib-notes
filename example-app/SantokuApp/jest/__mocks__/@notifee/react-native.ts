import notifee from '@notifee/react-native';
export * from '@notifee/react-native/dist/types/Library';
export * from '@notifee/react-native/dist/types/Notification';
export * from '@notifee/react-native/dist/types/Trigger';
export * from '@notifee/react-native/dist/types/NotificationIOS';
export * from '@notifee/react-native/dist/types/NotificationAndroid';

const mock: typeof notifee = {
  cancelAllNotifications: jest.fn(),
  cancelDisplayedNotifications: jest.fn(),
  cancelTriggerNotifications: jest.fn(),
  cancelNotification: jest.fn(),
  cancelDisplayedNotification: jest.fn(),
  cancelTriggerNotification: jest.fn(),
  createChannel: jest.fn(),
  createChannels: jest.fn(),
  createChannelGroup: jest.fn(),
  createChannelGroups: jest.fn(),
  deleteChannel: jest.fn(),
  deleteChannelGroup: jest.fn(),
  displayNotification: jest.fn(),
  openAlarmPermissionSettings: jest.fn(),
  createTriggerNotification: jest.fn(),
  getTriggerNotificationIds: jest.fn(),
  getDisplayedNotifications: jest.fn(),
  getTriggerNotifications: jest.fn(),
  getChannel: jest.fn(),
  isChannelCreated: jest.fn(),
  isChannelBlocked: jest.fn(),
  getChannels: jest.fn(),
  getChannelGroup: jest.fn(),
  getChannelGroups: jest.fn(),
  getInitialNotification: jest.fn(),
  onBackgroundEvent: jest.fn(),
  onForegroundEvent: jest.fn(),
  openNotificationSettings: jest.fn(),
  registerForegroundService: jest.fn(),
  stopForegroundService: jest.fn(),
  requestPermission: jest.fn(),
  setNotificationCategories: jest.fn(),
  getNotificationCategories: jest.fn(),
  getNotificationSettings: jest.fn(),
  getBadgeCount: jest.fn(),
  setBadgeCount: jest.fn(),
  incrementBadgeCount: jest.fn(),
  decrementBadgeCount: jest.fn(),
  openBatteryOptimizationSettings: jest.fn(),
  isBatteryOptimizationEnabled: jest.fn(),
  getPowerManagerInfo: jest.fn(),
  openPowerManagerSettings: jest.fn(),
  hideNotificationDrawer: jest.fn(),
  SDK_VERSION: '5.4.1',
};

// 複数のファイルでnotifeeをimportしていた場合に、redefineされないようにモックが存在するか確認
if (!__mocks.notifee) {
  Object.defineProperty(__mocks, 'notifee', {value: mock});
}

// テストケースごとにモックは初期化しておく。
beforeEach(() =>
  Object.values(mock).forEach(fn => {
    if (jest.isMockFunction(fn)) {
      fn.mockClear();
    }
  }),
);

export default mock;
