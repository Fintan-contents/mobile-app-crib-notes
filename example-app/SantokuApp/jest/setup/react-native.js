// https://reactnavigation.org/docs/testing#mocking-native-modules
// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

// !!! NOTICE !!!
// The React Navigation documentation says that `react-native-reanimated` and `react-native-gesture-handler`
// also need to be configured with mocks.
// But since React Navigation v6 does not already depend on those packages, mocking those packages is not necessary
// for React Navigation v6.
//
// If you want to use `react-native-reanimated` and `react-native-gesture-handler`, add mock according to the guide
// provided by each package.
// - `react-native-reanimated`: https://docs.swmansion.com/react-native-reanimated/docs/guide/testing/
// - `react-native-gesture-handler`: https://docs.swmansion.com/react-native-gesture-handler/docs/guides/testing
