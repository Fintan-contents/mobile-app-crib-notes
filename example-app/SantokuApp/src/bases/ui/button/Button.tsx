import React from 'react';
import {StyleSheet} from 'react-native';
import {Button as RNEButton, ButtonProps as RNEButtonProps} from 'react-native-elements';

const buttonWidthSizes = {
  small: 90,
  middle: 160,
  large: 260,
  full: '100%',
} as const;

type ButtonSizeType = keyof typeof buttonWidthSizes;

export type ButtonProps = RNEButtonProps & {
  /**
   * Button width size.
   */
  size?: ButtonSizeType;
};

export const Button: React.FC<ButtonProps> = props => {
  const {size, ...buttonProps} = props;
  const widthStyle = {width: buttonWidthSizes[size ?? 'small']};
  const buttonStyle = StyleSheet.flatten([styles.button, widthStyle, buttonProps.buttonStyle]);
  const containerStyle = StyleSheet.flatten([styles.container, widthStyle, buttonProps.containerStyle]);

  return <RNEButton {...buttonProps} buttonStyle={buttonStyle} containerStyle={containerStyle} />;
};

Button.defaultProps = {
  raised: true,
};

const styles = StyleSheet.create({
  button: {
    borderRadius: 8,
    height: 50,
  },
  container: {
    borderRadius: 8,
  },
});
