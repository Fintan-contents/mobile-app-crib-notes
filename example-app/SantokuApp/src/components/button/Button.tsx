import React from 'react';
import {StyleSheet} from 'react-native';
import {Button as RNEButton, ButtonProps} from 'react-native-elements';

const buttonWidthSizes = {
  small: 90,
  middle: 140,
  large: 200,
  full: '100%',
} as const;

type ButtonSizeType = keyof typeof buttonWidthSizes;

type Props = ButtonProps & {
  size?: ButtonSizeType;
};

export const Button: React.FC<Props> = (props) => {
  const {size, ...buttonProps} = props;
  const widthStyle = {width: buttonWidthSizes[size ?? 'small']};
  const buttonStyle = StyleSheet.flatten([styles.button, widthStyle]);
  const containerStyle = StyleSheet.flatten([styles.container, widthStyle]);

  return <RNEButton {...buttonProps} buttonStyle={buttonStyle} containerStyle={containerStyle} />;
};

Button.defaultProps = {
  raised: true,
};

const styles = StyleSheet.create({
  button: {
    borderRadius: 8,
    height: 50,
  },
  container: {
    borderRadius: 8,
  },
});
