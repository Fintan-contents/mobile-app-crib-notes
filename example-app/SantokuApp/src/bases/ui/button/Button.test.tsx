/**
 * Copyright 2023 TIS Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {fireEvent, render, screen} from '@testing-library/react-native';
import React from 'react';

import {Button} from './Button';

describe('Button', () => {
  it('Buttonが正常にrenderできることを確認', () => {
    render(<Button title="ボタン" />);
    expect(screen.queryByText('ボタン')).not.toBeNull();
    expect(screen).toMatchSnapshot();
  });

  it('Buttonのプレス時にコールバックが実行されることを確認', () => {
    const handlePress = jest.fn();
    render(<Button testID="button" onPress={handlePress} />);
    fireEvent.press(screen.getByTestId('button'));
    expect(handlePress).toHaveBeenCalledTimes(1);
  });

  it('Buttonのサイズ指定がrenderに反映されることを確認', () => {
    // スナップショットに反映されているか確認
    expect(render(<Button title="ボタン" size="small" />)).toMatchSnapshot('size small');
    expect(render(<Button title="ボタン" size="middle" />)).toMatchSnapshot('size middle');
    expect(render(<Button title="ボタン" size="large" />)).toMatchSnapshot('size large');
    expect(render(<Button title="ボタン" size="full" />)).toMatchSnapshot('size full');
  });

  it('Buttonのprops（buttonStyle / containerStyle）が指定通りrenderに反映されることを確認', () => {
    // スナップショットに反映されているか確認
    expect(
      render(
        <Button
          title="ボタン"
          size="small"
          buttonStyle={{backgroundColor: 'red', width: 150}}
          containerStyle={{width: 200}}
        />,
      ),
    ).toMatchSnapshot();
  });
});
