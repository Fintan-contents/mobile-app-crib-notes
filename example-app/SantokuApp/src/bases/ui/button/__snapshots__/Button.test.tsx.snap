// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Buttonが正常にrenderできることを確認 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 3,
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 3,
      },
      Object {
        "borderRadius": 8,
        "width": 90,
      },
      Object {
        "backgroundColor": "#fff",
        "overflow": "visible",
        "shadowColor": "rgba(0,0,0, .4)",
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": false,
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderColor": "#2089dc",
          "borderRadius": 8,
          "borderWidth": 0,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "padding": 8,
          "width": 90,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "paddingVertical": 1,
            "textAlign": "center",
          }
        }
      >
        ボタン
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button Buttonのprops（buttonStyle / containerStyle）が指定通りrenderに反映されることを確認 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 3,
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 3,
      },
      Object {
        "borderRadius": 8,
        "width": 200,
      },
      Object {
        "backgroundColor": "#fff",
        "overflow": "visible",
        "shadowColor": "rgba(0,0,0, .4)",
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": false,
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "red",
          "borderColor": "#2089dc",
          "borderRadius": 8,
          "borderWidth": 0,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "padding": 8,
          "width": 150,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "paddingVertical": 1,
            "textAlign": "center",
          }
        }
      >
        ボタン
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button Buttonのサイズ指定がrenderに反映されることを確認: size full 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 3,
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 3,
      },
      Object {
        "borderRadius": 8,
        "width": "100%",
      },
      Object {
        "backgroundColor": "#fff",
        "overflow": "visible",
        "shadowColor": "rgba(0,0,0, .4)",
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": false,
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderColor": "#2089dc",
          "borderRadius": 8,
          "borderWidth": 0,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "padding": 8,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "paddingVertical": 1,
            "textAlign": "center",
          }
        }
      >
        ボタン
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button Buttonのサイズ指定がrenderに反映されることを確認: size large 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 3,
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 3,
      },
      Object {
        "borderRadius": 8,
        "width": 260,
      },
      Object {
        "backgroundColor": "#fff",
        "overflow": "visible",
        "shadowColor": "rgba(0,0,0, .4)",
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": false,
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderColor": "#2089dc",
          "borderRadius": 8,
          "borderWidth": 0,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "padding": 8,
          "width": 260,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "paddingVertical": 1,
            "textAlign": "center",
          }
        }
      >
        ボタン
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button Buttonのサイズ指定がrenderに反映されることを確認: size middle 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 3,
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 3,
      },
      Object {
        "borderRadius": 8,
        "width": 160,
      },
      Object {
        "backgroundColor": "#fff",
        "overflow": "visible",
        "shadowColor": "rgba(0,0,0, .4)",
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": false,
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderColor": "#2089dc",
          "borderRadius": 8,
          "borderWidth": 0,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "padding": 8,
          "width": 160,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "paddingVertical": 1,
            "textAlign": "center",
          }
        }
      >
        ボタン
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button Buttonのサイズ指定がrenderに反映されることを確認: size small 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 3,
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 3,
      },
      Object {
        "borderRadius": 8,
        "width": 90,
      },
      Object {
        "backgroundColor": "#fff",
        "overflow": "visible",
        "shadowColor": "rgba(0,0,0, .4)",
        "shadowOffset": Object {
          "height": 1,
          "width": 1,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": false,
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderColor": "#2089dc",
          "borderRadius": 8,
          "borderWidth": 0,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "padding": 8,
          "width": 90,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "paddingVertical": 1,
            "textAlign": "center",
          }
        }
      >
        ボタン
      </Text>
    </View>
  </View>
</View>
`;
